### PR TITLE
[storage] Introduce `Archive`

### PR DESCRIPTION
**Pull request by patrick-ogrady**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new `Archive` module for a write-once key-value store optimized for high throughput and low-latency reads.
	- Added benchmarking capabilities for `HashMap` operations to evaluate performance.
	- New dependencies added to enhance functionality, including `prometheus-client` and `rand`.

- **Bug Fixes**
	- Enhanced error handling in the journal module for various scenarios.

- **Documentation**
	- Comprehensive documentation added for the `journal` module and the `Archive` module, including usage and error handling.

- **Tests**
	- Expanded test coverage for new features and error handling scenarios.

## TODO:
- [x] Support Blob cloning for concurrent access: #153 
- [x] Add tests for `storage::archive`
- [x] Add extensive documentation for both `Journal` and `Archive` (like `p2p::authenticated`): #160
- [x] Switch to lazy index pruning to reduce RAM usage: #158 
- [x] Don't allow pruning backwards (can't prune at a lower index): #158 
- [x] Avoid reading entire journal on replay with skip reads: #159 
- [x] Handle byte corruption differently than truncation (don't try to fix): #159
- [x] Separate metric for unnecessary prefix reads vs full reads
- [x] Consider removing padding from `Index` (u32 requires 4 extra bytes of padding) -> could store more of the key to minimize extra reads or change to u32 offset: #161 
- [x] Add test for approaching edge of max journal size: #161 
- [x] Add note that "alive" sections result in one open blob (i.e. file): #161 
- [x] Add compression support (like in `p2p`)

### CODE REVIEW

**Comment by patrick-ogrady** [comment]
@coderabbitai review

File: storage/src/archive/benches/hashmap_iteration.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+use criterion::{black_box, criterion_group, criterion_main, BatchSize, Criterion};
+use rand::Rng;
+use std::collections::HashMap;
+
+struct MockIndex {
+    section: u64,
+    _offset: usize,
+    _next: Option<Box<MockIndex>>,
+}
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/benches/hashmap_iteration.rs]: This is purposely kept there to approximate the behavior of the internal `Index` struct in `Archive`.

File: storage/src/archive/benches/hashmap_iteration.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+}
+
+criterion_group! {
+    name = benches;
+    config = Criterion::default().sample_size(10);
+    targets = benchmark_hashmap_iteration
+}
+criterion_main!(benches);
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/benches/hashmap_iteration.rs:52]: This is done because some samples take so long.

File: storage/src/archive/benches/hashmap_load_fixed.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                    for (key, section, offset) in v {
+                        let value = MockIndex {
+                            _section: section,
+                            _offset: offset,
+                            _next: None,
+                        };
+                        map.insert(key, value);
+                    }
+                },
+                BatchSize::SmallInput,
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/benches/hashmap_load_fixed.rs]: This is purposely not done because we are benchmarking additions when the size is not known.

File: storage/src/archive/benches/hashmap_load_fixed.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+                    // Populate vec with dummy data
+                    for _ in 0..n {
+                        let key: [u8; 4] = rng.gen();
+                        let section = rng.gen();
+                        let offset = rng.gen();
+                        vec.push((key, section, offset));
+                    }
+                    vec
+                },
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/benches/hashmap_load_fixed.rs]: I'm only interested in testing one configuration now, but if that changes this would be the right way to support it.

File: storage/src/archive/translator.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        assert_eq!(
+            translator.transform(b"abcdefgh"),
+            [b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h']
+        );
+        assert_eq!(
+            translator.transform(b"abcdefghijk"),
+            [b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h']
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/translator.rs:72]: I find these tests sufficient.

File: storage/src/archive/translator.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#[derive(Clone)]
+pub struct EightCap;
+
+impl Translator for EightCap {
+    type Key = [u8; 8];
+
+    fn transform(&self, key: &[u8]) -> Self::Key {
+        cap(key)
+    }
+}
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/translator.rs]: Nice!

File: storage/src/archive/storage.rs [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+            } else {
+                break true;
+            }
+        };
+
+        // If there is no valid head, remove key
+        if !found {
+            self.keys.remove(index_key);
+            return Ok(());
+        }
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/storage.rs]: Agreed!

File: storage/src/archive/storage.rs [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        force_sync: bool,
+    ) -> Result<(), Error> {
+        // Check key length
+        if key.len() != self.cfg.key_len {
+            return Err(Error::InvalidKeyLength);
+        }
+
+        // Check last pruned
+        let oldest_allowed = self.oldest_allowed.unwrap_or(0);
+        if section < oldest_allowed {
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/storage.rs:283]: This is intended.

File: storage/src/archive/storage.rs [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        }
+
+        // Update pending writes
+        let pending_writes = self.pending_writes.entry(section).or_default();
+        *pending_writes += 1;
+        if *pending_writes > self.cfg.pending_writes || force_sync {
+            self.journal.sync(section).await.map_err(Error::Journal)?;
+            *pending_writes = 0;
+            self.syncs.inc();
+        }
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/storage.rs:344]: Good catch!

File: storage/src/journal/storage.rs [hunk 10]
```diff
@@ -47,133 +49,260 @@
...
+        // Compute next offset
+        let offset = offset
+            .checked_add(4)
+            .ok_or(Error::OffsetOverflow)?
+            .checked_add(size)
+            .ok_or(Error::OffsetOverflow)?
+            .checked_add(4)
+            .ok_or(Error::OffsetOverflow)?;
+        Ok((offset, size, item_prefix))
+    }
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/journal/storage.rs]: This is expected and warned against in the docs.

File: runtime/src/tokio.rs [hunk 11]
```diff
@@ -501,23 +504,51 @@ pub struct Blob {
...
+    ) -> Self {
         metrics.open_blobs.inc();
         Self {
             metrics,
             partition,
             name: name.into(),
-            file,
+            file: Arc::new(AsyncMutex::new((file, len))),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply runtime/src/tokio.rs:531]: This is required because tokio requires a file to be seeked prior to access. When this is replaced with `mmap'ing`, we can make this much more efficient: #154

File: runtime/src/tokio.rs [hunk 12]
```diff
@@ -594,53 +637,82 @@
 
 impl crate::Blob for Blob {
     async fn len(&self) -> Result<usize, Error> {
-        let metadata = self.file.metadata().await.map_err(|_| Error::ReadFailed)?;
-        let len = metadata.len() as usize;
-        Ok(len)
+        let (_, len) = *self.file.lock().await;
+        Ok(len as usize)
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply runtime/src/tokio.rs]: This is not an option for this variable.

File: storage/src/journal/mod.rs [hunk 13]
```diff
@@ -24,25 +86,37 @@
...
+    #[error("unexpected size: expected={0} actual={1}")]
+    UnexpectedSize(usize, usize),
 }
 
 /// Configuration for `journal` storage.
 #[derive(Clone)]
 pub struct Config {
+    /// Registry for metrics.
+    pub registry: Arc<Mutex<Registry>>,
+
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/journal/mod.rs:136]: This is wrong. The caller must have mutable access to `Registry` to register metrics

File: storage/src/archive/mod.rs [hunk 14]
```diff
@@ -0,0 +1,785 @@
...
+
+use prometheus_client::registry::Registry;
+use std::{
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+use thiserror::Error;
+
+/// Errors that can occur when interacting with the archive.
+#[derive(Debug, Error)]
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply storage/src/archive/mod.rs:142]: Good catch.

File: runtime/src/deterministic.rs [hunk 15]
```diff
@@ -1023,6 +1034,8 @@
 }
 
 /// Implementation of [`crate::Blob`] for the `deterministic` runtime.
+///
+/// TODO: explain how design works with clones.
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply runtime/src/deterministic.rs]: Removing instead.

File: runtime/src/deterministic.rs [hunk 16]
```diff
@@ -1143,26 +1169,30 @@
...
         let partition = partitions
             .get_mut(&self.partition)
             .ok_or(Error::PartitionMissing(self.partition.clone()))?;
-        let blob = partition
+        let content = partition
             .blobs
             .get_mut(&self.name)
             .ok_or(Error::BlobMissing(self.partition.clone(), hex(&self.name)))?;
-        blob.clone_from(&self.content);
+        content.clone_from(&*self.content.read().unwrap());
```
[the remark this answers is not in the document]
Comment (patrick-ogrady) [reply runtime/src/deterministic.rs]: Agreed.
